src/game/suggest.go

```diff
@@ -138,3 +155,104 @@ func SolvePuzzle(shuffledGrid Grid, solvedGrid Grid, timeout chan bool) ([]Coord
 	}
 	return make([]Coords, 0), errors.New("No solution found by the deep puzzle algorithm")
 }
+
+func BuildNeighborList(node Node, solvedGrid Grid, possibleMoves []Coords) []Node {
+	var subList []Node
+
+	for _, coords := range possibleMoves {
+		neighborGrid, _ := Move(node.Grid, coords)
+		heuristic := TaxicabWithValues(neighborGrid, solvedGrid)
+		neighborNode := Node{node.Cost + 1 + heuristic, heuristic, neighborGrid, append(node.Moves, coords)}
+		subList = append(subList, neighborNode)
+	}
+	return subList
+}
+
+func BuildDepthList(list []Node, solvedGrid Grid, listChan chan []Node, timeoutChan chan bool) {
+	var sublist []Node
+
+	for _, node := range list {
+		select {
+		case <-timeoutChan:
+			return
+		default:
+			previousMoveCoords := node.Moves[len(node.Moves)-1]
+			previousValue := node.Grid[previousMoveCoords.Y][previousMoveCoords.X]
+			possibleMoves, _ := ListMovableTilesWithoutGoingBack(node.Grid, previousValue)
+			for _, neighbor := range BuildNeighborList(node, solvedGrid, possibleMoves) {
+				sublist = append(sublist, neighbor)
+			}
+		}
+	}
+	listChan <- sublist
+}
+
+func findGridResolvedInList(list []Node) (Node, error) {
+	var node Node
+	for _, node = range list {
+		if node.Heuristic == 0 {
+			return node, nil
+		}
+	}
+	return node, errors.New("No solved grid found in list")
+}
+
+func SolvePuzzleD(shuffledGrid Grid, solvedGrid Grid) ([]Coords, error) {
+	possibleMoves, _ := ListMovableTiles(shuffledGrid)
+	return solvePuzzleD(shuffledGrid, solvedGrid, possibleMoves)
+}
+
+func SolvePuzzleDWithPreviousMove(shuffledGrid Grid, solvedGrid Grid, lastMove Coords) ([]Coords, error) {
+	previousValue := shuffledGrid[lastMove.Y][lastMove.X]
+	possibleMoves, _ := ListMovableTilesWithoutGoingBack(shuffledGrid, previousValue)
+
+	return solvePuzzleD(shuffledGrid, solvedGrid, possibleMoves)
+}
+
+func solvePuzzleD(shuffledGrid Grid, solvedGrid Grid, possibleMoves []Coords) ([]Coords, error) {
```

Comment:
Can you try to split this function to make it clearer?

Reply:
Moreover it'll help you to test because this function is complex (and important)

Reply:
I've split a little bit this function. I think the whole file needs to be split in a specific submodule to distinct the heuristic, the priority list etc.